fuel-vm/src/interpreter/alu.rs

```diff
@@ -13,79 +13,161 @@ where
     where
         F: FnOnce(B, C) -> (u128, bool),
     {
-        Self::is_register_writable(ra)?;
-
-        let (result, _overflow) = f(b, c);
-
-        if result > Word::MAX as u128 && !self.is_wrapping() {
-            return Err(PanicReason::ArithmeticOverflow.into());
-        }
-
-        // set the OF register to high bits of the u128 result
-        self.registers[RegId::OF] = (result >> 64) as u64;
-        self.registers[RegId::ERR] = 0;
-
-        // set the return value to the low bits of the u128 result
-        self.registers[ra] = (result & Word::MAX as u128) as u64;
-
-        self.inc_pc()
+        let (ReadRegisters { flag, of, err, pc, .. }, mut w) = split_registers(&mut self.registers);
+        let dest = &mut w[ra.try_into()?];
```

Comment:
It'll be difficult to land that while these big PRs are in flight, unless we want to make a PR that targets this one directly.

Reply:
s is actually checking that the key is above the system registers and below the end of all registers. I don't think `RegKey` does that?

Reply:
This is a really nice way to prevent invalid register ids from being written to! We had this bug before and it was discovered via proptesting, but having a way to guarantee the correctness via the type system is much better.